src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -929,125 +725,316 @@ internal ThreadLocalList(int ownerThreadId)
+                        _currentOp = (int)ConcurrentBagListOperation.None; // set back to None to avoid a deadlock with steals
+                        Monitor.Enter(this, ref lockTaken);
+                        if (_headIndex < tail)
+                        {
+                            // Element still available. Peek it.
+                            result = _array[(tail - 1) & _mask];
+                            return true;
+                        }
+                        else
+                        {
+                            // We encountered a race condition and the element was stolen, nothing to peek.
+                            result = default(T);
+                            return false;
+                        }
+                    }
+                }
+                finally
                 {
-                    result = head._value;
-                    return true;
+                    _currentOp = (int)ConcurrentBagListOperation.None;
+                    if (lockTaken)
+                    {
+                        Monitor.Exit(this);
+                    }
                 }
-                result = default(T);
-                return false;
             }
 
-            /// <summary>
-            /// Steal an item from the tail of the list
-            /// </summary>
+            /// <summary>Steal an item from the tail of the list.</summary>
             /// <param name="result">the removed item</param>
             /// <param name="remove">remove or peek flag</param>
-            internal void Steal(out T result, bool remove)
+            internal bool TrySteal(out T result, bool remove)
             {
-                Node tail = _tail;
-                Debug.Assert(tail != null);
-                if (remove) // Take operation
+                while (true)
                 {
-                    _tail = _tail._prev;
-                    if (_tail != null)
+                    // Fast-path check to see if the list is empty.
+                    if (_headIndex >= _tailIndex)
                     {
-                        _tail._next = null;
+                        result = default(T);
+                        return false;
                     }
-                    else
+
+                    // Anything else requires synchronization.
+                    lock (this)
                     {
-                        _head = null;
+                        // Wait for thread-local operations on the list to subside, if there are
+                        // few enough elements in the list that we could conflict with such operations.
+                        if (Count <= 2)
+                        {
+                            if (_currentOp != (int)ConcurrentBagListOperation.None)
+                            {
+                                var spinner = new SpinWait();
+                                while (_currentOp != (int)ConcurrentBagListOperation.None)
```

Comment:
Ah, it appears to be necessary, but only for peeks.  Peeks are special, in that they need to be able to read the value without taking it or preventing someone else from taking it, which means they can't mess with the head/tail pointers in a way that would prevent a steal from getting the item.  As such, a peek needs to set this flag, so that if the peek contends with a steal, the steal waits for the peek to finish; otherwise, the peek can end up reading a default(T) or a torn T.  I can change the condition here to old spin for peeks.  It's not necessary to do so for pops or pushes, as the logic for handling contention with those is already built in to the rest of the logic.